src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,122 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
     }
   }
 
+  // Convert a + a + ... + a into a*n
+  Node* serial_additions = convert_serial_additions(phase, bt);
+  if (serial_additions != nullptr) {
+    return serial_additions;
+  }
+
   return AddNode::Ideal(phase, can_reshape);
 }
 
+// Try to convert a serial of additions into a single multiplication. Also convert `(a * CON) + a` to `(CON + 1) * a` as
+// a side effect. On success, a new MulNode is returned.
+Node* AddNode::convert_serial_additions(PhaseGVN* phase, BasicType bt) {
+  // We need to make sure that the current AddNode is not part of a MulNode that has already been optimized to a
+  // power-of-2 addition (e.g., 3 * a => (a << 2) + a). Without this check, GVN would keep trying to optimize the same
+  // node and can't progress. For example, 3 * a => (a << 2) + a => 3 * a => (a << 2) + a => ...
+  if (find_power_of_two_addition_pattern(this, bt).valid) {
+    return nullptr;
+  }
+
+  Node* in1 = in(1);
+  Node* in2 = in(2);
+
+  // While multiplications can be potentially optimized to power-of-2 subtractions (e.g., a * 7 => (a << 3) - a),
+  // (x - y) + y => x is already handled by the Identity() methods. So, we don't need to check for that pattern here.
+  Multiplication mul;
+  if (
+      ((mul = find_simple_addition_pattern(in1, bt)).valid && mul.variable == in2) ||
+      ((mul = find_simple_lshift_pattern(in1, bt)).valid && mul.variable == in2) ||
+      ((mul = find_simple_multiplication_pattern(in1, bt)).valid && mul.variable == in2) ||
+      ((mul = find_power_of_two_addition_pattern(in1, bt)).valid && mul.variable == in2)
+      ) {
+    Node* con = (bt == T_INT)
+                ? (Node*) phase->intcon((jint) (mul.multiplier + 1)) // intentional type narrowing to allow overflow at max_jint
+                : (Node*) phase->longcon((mul.multiplier + 1));
```

Comment:
I think just to be safe, you should use `java_add` to have correct overflow semantics. You are using `jlong` for `multiplier`, which is a signed integer type, and in C++ overflow is undefined behavior as far as I know, let's avoid that ;)

Actually, do you have a test where the multiplier overflows here?

Reply:
Good point regarding UB. Updated to use `java_add`.

There are several test cases named `.*Overflow()` in `TestSerialAdditions`.